src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RandomNumberGenerator.cs

```diff
@@ -163,6 +164,200 @@ public static byte[] GetBytes(int count)
+        }
+
+        /// <summary>
+        ///   Performs an in-place shuffle of a span using cryptographically random number generation.
+        /// </summary>
+        /// <param name="values">The span to shuffle.</param>
+        /// <typeparam name="T">The type of span.</typeparam>
+        public static void Shuffle<T>(Span<T> values)
+        {
+            int n = values.Length;
+
+            for (int i = 0; i < n - 1; i++)
+            {
+                int j = GetInt32(i, n);
+
+                if (i != j)
+                {
+                    T temp = values[i];
+                    values[i] = values[j];
+                    values[j] = temp;
+                }
+            }
+        }
+
+        private static void GetHexStringCore(Span<char> destination, bool lowercase)
+        {
+            Debug.Assert(!destination.IsEmpty);
+
+            const int RandomBufferSize = 64; // If this changes, the tests need to be updated since they try to exercise boundary conditions.
+            Span<byte> randomBuffer = stackalloc byte[RandomBufferSize];
+            HexConverter.Casing casing = lowercase ? HexConverter.Casing.Lower : HexConverter.Casing.Upper;
+
+            // Don't overfill the buffer if the destination is smaller than the buffer size. We need to round up when
+            // when dividing by two to account for an odd-length destination.
+            int needed = (destination.Length + 1) / 2;
+            Span<byte> remainingRandom = randomBuffer.Slice(0, Math.Min(RandomBufferSize, needed));
+            RandomNumberGenerator.Fill(remainingRandom);
+
+            // HexConverter can only write in multiples of two. If the length is odd, get back to an even length.
+            if (destination.Length % 2 != 0)
+            {
+                destination[0] = lowercase ?
+                    HexConverter.ToCharLower(remainingRandom[0]) :
+                    HexConverter.ToCharUpper(remainingRandom[0]);
+
+                destination = destination.Slice(1);
+                remainingRandom = remainingRandom.Slice(1);
+            }
+
+            while (!destination.IsEmpty)
+            {
+                needed = destination.Length / 2;
+
+                if (remainingRandom.IsEmpty)
+                {
+                    remainingRandom = randomBuffer.Slice(0, Math.Min(RandomBufferSize, needed));
+                    RandomNumberGenerator.Fill(remainingRandom);
+                }
+
+                HexConverter.EncodeToUtf16(remainingRandom, destination, casing);
+                destination = destination.Slice(remainingRandom.Length * 2);
+                remainingRandom = default;
+            }
+        }
+
+        private static void GetItemsCore<T>(ReadOnlySpan<T> choices, Span<T> destination)
+        {
+            for (int i = 0; i < destination.Length; i++)
+            {
+                destination[i] = choices[GetInt32(choices.Length)];
+            }
```

Comment:
Similar to the N virtual calls concern in System.Random, it might be useful to change this to do something like

```C#
Span<byte> randomBytes = stackalloc byte[128];
Span<int> randomIntStorage = stackalloc int[randomBytes.Length / sizeof(int)];
int randomInts = 0;

for (int i = destination.Length - 1; i >= 0; i--)
{
    while (randomInts <= 0)
    {
        if (i < randomIntStorage.Length)
        {
            randomBytes = randomBytes.Slice(0, i * sizeof(int));
        }

        Fill(randomBytes);
        randomInts = FilterInt32(randomBytes, randomIntStorage);
    }

    randomInts--;
    destination[i] = choices[randomInts[randomInts]];
}
```

to reduce P/Invokes by a factor of randomIntStorage.Length.

Though that requires writing FilterInt32, and the complexity of above, so it should be checked to see if the perf justifies it.